From 19471bda3d488ff5ccf53589421a1a4254735855 Mon Sep 17 00:00:00 2001
From: Eyal Shapira <eyal@wizery.com>
Date: Thu, 13 Sep 2012 13:53:00 +0300
Subject: [PATCH 26/40] mac80211: fix sync of rx filters with driver and cfg80211

Avoid changing the local saved rx filters in case
the driver fails to set them. As the driver clears
the filters whenever the device is stopped we need
to configure them back when turning on the device to
avoid being out of sync with cfg80211 which reports
back to userspace what's configured.
Also fix missing locking in ieee80211_set_rx_filters.

Signed-off-by: Eyal Shapira <eyal@wizery.com>
---
 drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/cfg.c   |   10 +++++++++-
 drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/iface.c |    1 +
 drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/util.c  |    2 +-
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/cfg.c b/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/cfg.c
index 679f844..76a2cba 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/cfg.c
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/cfg.c
@@ -1907,12 +1907,20 @@ int ieee80211_set_rx_filters(struct wiphy *wiphy,
 			     struct cfg80211_wowlan *wowlan)
 {
 	struct ieee80211_local *local = wiphy_priv(wiphy);
+	int ret = 0;
 
 	if (!(local->hw.flags & IEEE80211_HW_SUPPORTS_RX_FILTERS))
 		return 0;
 
+	mutex_lock(&local->mtx);
+	ret = drv_set_rx_filters(local, wowlan);
+	if (ret < 0)
+		goto unlock;
+
 	local->wowlan_patterns = wowlan;
-	return drv_set_rx_filters(local, wowlan);
+unlock:
+	mutex_unlock(&local->mtx);
+	return ret;
 }
 
 static int ieee80211_set_cqm_rssi_config(struct wiphy *wiphy,
diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/iface.c b/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/iface.c
index 6b51f8b..632bf81 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/iface.c
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/iface.c
@@ -279,6 +279,7 @@ static int ieee80211_do_open(struct net_device *dev, bool coming_up)
 			res = drv_add_interface(local, sdata);
 			if (res)
 				goto err_stop;
+			drv_set_rx_filters(local, local->wowlan_patterns);
 		}
 
 		if (sdata->vif.type == NL80211_IFTYPE_AP) {
diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/util.c b/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/util.c
index 1c86742..3a11c8c 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/util.c
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/net/mac80211/util.c
@@ -1220,7 +1220,7 @@ int ieee80211_reconfig(struct ieee80211_local *local)
 	ieee80211_hw_config(local, ~0);
 
 	ieee80211_configure_filter(local);
-	ieee80211_set_rx_filters(hw->wiphy, local->wowlan_patterns);
+	drv_set_rx_filters(local, local->wowlan_patterns);
 
 	/* Finally also reconfigure all the BSS information */
 	list_for_each_entry(sdata, &local->interfaces, list) {
-- 
1.7.0.4

